orchestra/tests/test_notifications.py

```diff
@@ -219,3 +220,49 @@ def _validate_slack_messages(message_stub):
             _validate_slack_messages('Task has been aborted.')
         self.assertEquals(len(internal_slack_messages), 0)
         self.assertEquals(len(experts_slack_messages), 0)
+
+    def test_slack_group_id(self):
+        def fake_random_string():
+            fake_random_string.num_calls = (
+                getattr(fake_random_string, 'num_calls', 0) + 1)
+            return str(fake_random_string.num_calls)
+
+        for project_id, group_name in (('empty_project', 'ketchup-3-sales-1'),
+                                       ('base_test_project',
+                                        'bongo-bash723581-3'),
+                                       ('no_task_assignments',
+                                        'bongo-bash723581-4')):
+            project = self.projects[project_id]
+            project.slack_group_id = group_name
+            project.save()
+
+        project = self.projects['reject_rev_proj']
+
+        # Ensure slack group names are properly slugified and randomized.
+        project.short_description = "Sally's Plumbing Emporium | Write Copy"
+        group_name = _project_slack_group_name(project)
+        self.assertTrue(group_name.startswith('sallys-plumbing'))
+        self.assertEqual(len(group_name), 20)
+
+        # Ensure slack group names are properly slugified and randomized.
+        project.short_description = "Sally's Plumbingorama Emporium | Write"
+        group_name = _project_slack_group_name(project)
+        self.assertTrue(group_name.startswith('sallys-plumbingo'))
+        self.assertEqual(len(group_name), 21)
+
+        # Test that we create unique slack IDs if there are conflicts by
+        # mocking the randomization logic to return deterministic results and
+        # ensure that we don't repeat project IDs.
+        with patch('orchestra.slack._random_string', new=fake_random_string):
+            for short_description, group_id in (
+                    ('ketchup #3 Sales --:/ kitchen supplies',
+                     'ketchup-3-sales-2'),
+                    ('!!bongo bash*&#$(*$7235818923701842312',
+                     'bongo-bash723581-5'),
```

Comment:
maybe this needs explicit comments of the form "-5 will be used, because the mock function's counter is at 3 and -3 and -4 already exist" -- I had to stare at this for a while before I got it.